src/optionals/path.c

```diff
@@ -233,6 +233,51 @@ static Value listDirNative(DictuVM *vm, int argCount, Value *args) {
     return OBJ_VAL(dir_contents);
 }
 
+static Value joinNative(DictuVM *vm, int argCount, Value *args) {
+    char* argCountError = "join() requires 1 or more arguments (%d given).";
+    char* nonStringError = "join() argument at index %d is not a string";
+
+    if (argCount == 1 && IS_LIST(args[0])) {
+        argCountError = "List passed to join() must have 1 or more elements (%d given).";
+        nonStringError = "The element at index %d of the list passed to join() is not a string";
+        ObjList *list = AS_LIST(args[0]);
+        argCount = list->values.count;
+        args = list->values.values;
+    }
+
+    if (argCount == 0) {
+        runtimeError(vm, argCountError, argCount);
+        return EMPTY_VAL;
+    }
+
+    for (int i = 0; i < argCount; ++i) {
+        if (!IS_STRING(args[i])) {
+            runtimeError(vm, nonStringError, i);
+            return EMPTY_VAL;
+        }
+    }
+
+    // resultSize = # of dir separators that will be used + length of each string arg
+    size_t resultSize = abs(argCount - 1); // abs is needed her because of a clang bug
+    for (int i = 0; i < argCount; ++i) resultSize += AS_STRING(args[i])->length;
+    // It's possible for resultSize to be too large if the strings already end with the dir
+    // separator, but having likely at most a few bytes of unused memory in the resulting string
+    // object's char array should be fine.
+
+    char* str = calloc(resultSize, sizeof *str);
```

Comment:
When allocating memory it's best to try and use the macros available to us. This will 1, cause a garbage collection to occur if required and 2. allow us to swap out the memory allocation scheme at a later date if ever required much easier.

```suggestion
    char* str = ALLOCATE(vm, char*, resultSize);
```